target/linux/ipq806x/base-files/lib/upgrade/buffalo.sh

```diff
@@ -0,0 +1,182 @@
+		echo "cannot find first ubi mtd partition $CI_BUF_UBIPART"
+		return 1
+	fi
+
+	# search second ubi partition
+	local mtdnum2="$( find_mtd_index "$CI_BUF_UBIPART2" )"
+	if [ ! "$mtdnum2" ]; then
+		echo "cannot find second ubi mtd partition $CI_BUF_UBIPART2"
+	fi
+
+	local ubidev="$( nand_find_ubi "$CI_BUF_UBIPART" )"
+	if [ ! "$ubidev" ]; then
+		ubiattach -m "$mtdnum"
+		sync
+		ubidev="$( nand_find_ubi "$CI_BUF_UBIPART" )"
+	fi
+
+	local ubidev2="$( nand_find_ubi "$CI_BUF_UBIPART2" )"
+	if [ ! "$ubidev2" ] && [ -n "$mtdnum2" ]; then
+		ubiattach -m "$mtdnum2"
+		sync
+		ubidev2="$( nand_find_ubi "$CI_BUF_UBIPART2" )"
+	fi
+
+	if [ ! "$ubidev" ]; then
+		ubiformat /dev/mtd$mtdnum -y
+		ubiattach -m "$mtdnum"
+		sync
+		ubidev="$( nand_find_ubi "$CI_BUF_UBIPART" )"
+	fi
+
+	local kern_ubivol="$( nand_find_volume $ubidev $KERN_VOLNAME )"
+	local root_ubivol="$( nand_find_volume $ubidev rootfs )"
+	local data_ubivol="$( nand_find_volume $ubidev rootfs_data )"
+	# Buffalo volumes
+	local fkroot_ubivol="$( nand_find_volume $ubidev $FKROOT_VOLNAME )"
+	local buf_data_ubivol="$( nand_find_volume $ubidev ubi_rootfs_data )"
+	local buf_kern2_ubivol="$( nand_find_volume $ubidev2 $KERN_VOLNAME )"
+
+	# remove ubiblock device of rootfs
+	local root_ubiblk="ubiblock${root_ubivol:3}"
+	if [ "$root_ubivol" -a -e "/dev/$root_ubiblk" ]; then
+		echo "removing $root_ubiblk"
+		if ! ubiblock -r /dev/$root_ubivol; then
+			echo "cannot remove $root_ubiblk"
+			return 1;
+		fi
+	fi
+
+	# kill volumes
+	[ "$kern_ubivol" ] && ubirmvol /dev/$ubidev -N $KERN_VOLNAME || true
+	[ "$fkroot_ubivol" ] && ubirmvol /dev/$ubidev -N $FKROOT_VOLNAME || true
+	[ "$root_ubivol" ] && ubirmvol /dev/$ubidev -N rootfs || true
+	[ "$data_ubivol" ] && ubirmvol /dev/$ubidev -N rootfs_data || true
+	[ "$buf_data_ubivol" ] && ubirmvol /dev/$ubidev -N ubi_rootfs_data || true
+
+	if ! ubimkvol /dev/$ubidev -N $KERN_VOLNAME -s $kernel_length; then
+		echo "cannot create kernel volume"
+		return 1;
+	fi
+
+	# re-create fake-rootfs volume and write data from $FKROOT_BIN
+	local fkroot_length="$(wc -c $FKROOT_BIN | awk '{print $1}')"
+	if ! ubimkvol /dev/$ubidev -N $FKROOT_VOLNAME -s $fkroot_length; then
+		echo "cannot create fake-rootfs volume"
+		return 1;
+	else
+		fkroot_ubivol="$( nand_find_volume $ubidev $FKROOT_VOLNAME )"
+		echo "write fake-rootfs data to $fkroot_ubivol"
+		ubiupdatevol /dev/$fkroot_ubivol -s $fkroot_length $FKROOT_BIN
+	fi
+
+	if ! ubimkvol /dev/$ubidev -N rootfs -s $rootfs_length; then
+		echo "cannot create rootfs volume"
+		return 1;
+	fi
+
+	if ! ubimkvol /dev/$ubidev -N rootfs_data -m; then
+		echo "cannot initialize rootfs_data volume"
+		return 1
+	fi
+
+	# remove kernel volume from second ubi partition for U-Boot
+	# U-Boot writes os-image into Bank2 from Bank1
+	[ "$buf_kern2_ubivol" ] && ubirmvol /dev/$ubidev2 -N $KERN_VOLNAME || true
+
+	sync
+	return 0
+}
+
+create_fakeroot() {
+	local fkroot_dir="/tmp/fkroot"
+
+	if [ ! -x "/usr/sbin/mksquashfs" ]; then
```

Comment:
Would it be possible to move this check into the preparation stage, before the ubi volumes are all removed?